### PR TITLE
Biallelic filtering and bcftools comparison

### PR DESCRIPTION
**Pull request by briankirz**
biallelic.py filters out all non-biallelic snps from a gzipped vcf file and writes them to a new vcf file.

compare_vcf confirms its accuracy with another output file resulting from a bcftools call to do the same.

### CODE REVIEW

**Review by David-Peede — CHANGES_REQUESTED** [review]
- [x]  Add the `bcftools` command in a shell script and annotate
- [x]  Annotate every line of your python functions
- [x] Incorporate the `INFO` field's `VT` flag
- [x] Add a README.MD File briefly explaining (one sentence) what each script in this directory does

**Review by David-Peede — CHANGES_REQUESTED** [review]
- [x] make the README.MD formatted as a markdown file
- [x] write the bcf_biallelic.sh as shell file that you will submit to oscar (if you get stuck look at my example [here](https://github.com/David-Peede/BIOL1435/blob/main/OSCAR/get1KsampleFile.sh) or at the [documentation](https://docs.ccv.brown.edu/oscar/submitting-jobs/shared-machine))
- [x] look at my `tgp_site_info` function starting on line 200 from [here](https://github.com/David-Peede/targt_pipeline_vcf/blob/master/vcf_functions.py) and update your `biallelic.py` function for readability purposes (also once you open a file you don't need to open it again)
- [x] run all of this on chromosome 20 and if everything looks good then run on chromosme 1 (currently I believe bcftools and your parser will not produce identical VCFs) 
- [x] create a `run_me.sh` which shows how to run all of the code on oscar

**Comment by briankirz** [comment]
Question about SLURM jobs:
What I got from our meeting yesterday was that I was doing computational work locally, while I should have been shell files/SLURM scripts in order to submit the job to OSCAR.
My understanding is that this is done by including #SBATCH parameters in the header. However, in your get1KsampleFile.sh example, I'm not noticing any header parameters for input.
If I want to execute the same code on two different files (ie, chromosome 1 and chromosome 20), how would I indicate their different input files? Would I have to change the job manually each time I do it? I searched "input" in the full SLURM documentation but all I found was something related to `sattach` and I feel like I'm going down the wrong road again
Screen Shot 2022-04-14 at 2 56 10 PM
.

**Comment by briankirz** [comment]
I'm wondering about how to run input into the SLURM job.
I want to load bcftools and then execute the command on a specific file. Here's what I've got so far
Screen Shot 2022-04-14 at 3 15 42 PM
I found --input=<name> in the SLURM documentation but I'm not sure how to use the name of that file. Will I need to hardcode a different SLURM job for the same command being run on different files?

Maybe I could take a look at the more complex SLURM job list that you showed me during our last meeting in order to look for clues?

**Comment by briankirz** [comment]
"also once you open a file you don't need to open it again" as it relates to biallelic.py
Not sure what you mean by this

**Comment by David-Peede** [comment]
> 

Look back through my tutorial on OSCAR. The SLURM header is just to request resources NOT EXECUTE CODE. You are still writing in python but bcftools is a command line tool much like doing `cd` or `pwd`

**Comment by David-Peede** [comment]
> "also once you open a file you don't need to open it again" as it relates to biallelic.py Not sure what you mean by this

`with gzip.open(vcf_input, 'rb') as gz, open(vcf_output, 'w') as f:` you only need the first open

**Comment by briankirz** [comment]
When I compare vcfs in compare_vcf.py, should I be preparing for the headers of each file to not necessarily be the same length? This is taking longer to implement, as I see that the file created by a SLURM job has a different header than the one created by my python code.

It would be much easier to implement if both files were identical (including headers), but I'm not sure that there's a way to do this where I can assume that.

**Comment by David-Peede** [comment]
> When I compare vcfs in compare_vcf.py, should I be preparing for the headers of each file to not necessarily be the same length? This is taking longer to implement, as I see that the file created by a SLURM job has a different header than the one created by my python code.
> 
> It would be much easier to implement if both files were identical (including headers), but I'm not sure that there's a way to do this where I can assume that.

Just skip any line starting with a `##` or `#` and there shouldn't be any problems

**Comment by briankirz** [comment]
Two things:
1) including the -I (exclude indels) argument in the second command results in an output that is blank besides the SLURM header
Screen Shot 2022-04-15 at 6 28 10 PM

However, without the -I it works fine. I'll keep experimenting

File: kirz/vcf_processing/README.MD [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# Filtering chromosomes to include only biallelic SNPS
```
Comment (David-Peede) [inline kirz/vcf_processing/README.MD:1]: - [x]  brief description of the purpose of this repo
- [x]  brief description of the code scripts including input and output description 
- [x] brief description of the other directories and what they contain
- [x] step by step example of how to recreate the analyses

**Review by David-Peede — CHANGES_REQUESTED** [review]
clean up file names and code

File: kirz/vcf_processing/README.MD (hunk 1 above)
Comment (David-Peede) [reply kirz/vcf_processing/README.MD:1]: - [ ] fix bash one liners

File: kirz/vcf_processing/bcf_biallelic.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (David-Peede) [inline kirz/vcf_processing/bcf_biallelic.sh:1]: - [x] remove this file

**Review by David-Peede — CHANGES_REQUESTED** [review]
- [x] fix readme
- [ ] remove unnecessary files

**Review by David-Peede — APPROVED** [review]

**Pull request merged**